File: pyround1/112.路径总和.py
#
# @lc app=leetcode.cn id=112 lang=python
#
# [112] 路径总和
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def hasPathSum(self, root, sum):
        """
        :type root: TreeNode
        :type sum: int
        :rtype: bool
        """
        if not root: return False
        stack = [(root.val, root)]
        while stack:
            sum_, node = stack.pop(0)
            if node.right:
                stack.append((sum_ + node.right.val, node.right))
            if node.left:
                stack.append((sum_ + node.left.val, node.left))
            if not node.left and not node.right and sum_ == sum:
                return True
        return False
# @lc code=end

File: 112.路径总和.cpp
/*
 * @lc app=leetcode.cn id=112 lang=cpp
 *
 * [112] 路径总和
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool hasPathSum(TreeNode* root, int sum) {
        return helper(root, 0, sum);
    }
    bool helper(TreeNode*root, int target, int sum){
        if(!root)
            return false;
        else if(root and !root->left and !root->right)
            return target + root->val == sum;
        else{
            return helper(root->left, target + root->val) or helper(root->right, target + root->val);
        }

    }

};
// @lc code=end

